src/common/analog.rs

```diff
@@ -32,3 +34,86 @@ pub trait AnalogReader<Word> {
     fn read(&mut self) -> Result<Word, Self::Error>;
     fn name(&self) -> String;
 }
+
+#[derive(Debug)]
+pub(crate) struct AnalogReaderConfig {
+    pub(crate) name: &'static str,
+    pub(crate) pin: i32,
+}
+
+impl TryFrom<Kind> for AnalogReaderConfig {
+    type Error = AttributeError;
+    fn try_from(value: Kind) -> Result<Self, Self::Error> {
+        match value {
+            Kind::StructValueStatic(v) => {
+                if !v.contains_key("name") {
+                    return Err(AttributeError::KeyNotFound);
+                }
+                if !v.contains_key("pin") {
+                    return Err(AttributeError::KeyNotFound);
+                }
+                let name = v.get("name").unwrap().try_into()?;
+                let pin: i32 = v.get("pin").unwrap().try_into()?;
+                Ok(Self { name, pin })
+            }
+            _ => Err(AttributeError::ConversionImpossibleError),
+        }
+    }
+}
+
+impl TryFrom<&Kind> for AnalogReaderConfig {
+    type Error = AttributeError;
+    fn try_from(value: &Kind) -> Result<Self, Self::Error> {
+        match value {
+            Kind::StructValueStatic(v) => {
```

Comment:
The T / &T impl duplication is a nuisance. I assume that had you found a way to avoid needing to write it all you would have done so, but it does at least look like the body of the two functions could be lifted into a common function. Something like:

```
fn arc_try_from(value: &Kind) -> Result<AnalogReaderConfig, AttributeError> {
    match value { ... }
}

impl TryFrom<Kind> for AnalogReaderConfig {
    type Error = AttributeError;
    fn try_from(value: Kind) -> Result<Self, Self::Error> {
        return arc_try_from(&value);
    }
}

impl TryFrom<Kind> for AnalogReaderConfig {
    type Error = AttributeError;
    fn try_from(value: &Kind) -> Result<Self, Self::Error> {
        return arc_try_from(value);
    }
}

```

Reply:
I think we could use the [std](https://doc.rust-lang.org/std/index.html)::[borrow](https://doc.rust-lang.org/std/borrow/index.html)::[Borrow](https://doc.rust-lang.org/std/borrow/trait.Borrow.html#) trait I haven't played with this yet will had a ticket to track this